sdk/attestation/Azure.Security.Attestation/src/AttestationToken.cs

```diff
@@ -143,43 +202,390 @@ protected AttestationToken()
+            {
+                if (!ValidateCommonProperties(options))
+                {
+                    return false;
+                }
+                if (options.ValidationCallback != null)
+                {
+                    return options.ValidationCallback(this, null);
+                }
+                return true;
+            }
+
+            // This token is a secured attestation token. If the caller provided signing certificates, then
+            // they need to have provided at least one certificate.
+            if (attestationSigningCertificates != null)
+            {
+                Argument.AssertNotNull(attestationSigningCertificates[0], nameof(attestationSigningCertificates));
+            }
+
+            var possibleCertificates = await GetCandidateSigningCertificatesAsync(attestationSigningCertificates, cancellationToken).ConfigureAwait(false);
+            if (possibleCertificates.Length == 0)
+            {
+                throw new Exception($"Unable to find any certificates which can be used to validate the token.");
+            }
+
+            if (!ValidateTokenSignature(possibleCertificates))
+            {
+                throw new Exception($"Could not find a certificate which was used to sign the token.");
+            }
+
+            if (!ValidateCommonProperties(options))
+            {
+                return false;
+            }
+            if (options.ValidationCallback != null)
+            {
+                return options.ValidationCallback(this, SigningCertificate);
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Validate a JSON Web Token returned by the MAA.
+        /// <para/>
+        /// If the caller provides a set of signers, than that set of signers will be used as the complete set of candidates for signing.
+        /// If the caller does not provide a set of signers, then the <see cref="ValidateToken(TokenValidationOptions, IReadOnlyList{AttestationSigner}, CancellationToken)"/>
+        /// API will a set of callers derived from the contents of the attestation token.
+        /// </summary>
+        /// <param name="options">Options used while validating the attestation token.</param>
+        /// <param name="attestationSigningCertificates">Signing Certificates used to validate the token.</param>
+        /// <param name="cancellationToken">Token used to cancel this operation if necessary.</param>
+        /// <returns>true if the token was valid, false otherwise.</returns>
+        /// <exception cref="ArgumentNullException">Thrown if the signing certificates provided are invalid.</exception>
+        /// <exception cref="Exception">Thrown if validation fails.</exception>
+        public virtual bool ValidateToken(TokenValidationOptions options, IReadOnlyList<AttestationSigner> attestationSigningCertificates, CancellationToken cancellationToken = default)
```

Comment:
Most of the code in this method is shared with the Async version of this method.  Can the code be shared?

Reply:
That is not the pattern used by other SDK functions. @jsquire, @tg-msft: What is the convention for sync and async functions? The patterns covered in the training have dual implementations, would it be viable to have the sync version of the API simply call the async version, or should the two implementations be separate?